### PR TITLE
Parallelize histogram and MPH calculations

### LINKED ISSUE #58: Median pulse height count array return process is inherently slow
**Is your feature request related to a problem? Please describe.**

The pulse height extraction process has been the largest bottleneck in the watcher pipeline by miles. While the actual process is very similar to `.bin` file extraction, the main difference is the median pulse height extraction has to return three arrays back to Python. The conversion process between Cython and Python can be very slow if the arrays are large.

**Describe the solution you'd like**

A couple of options here:

1. Let's first see if there's any way we can speed up the return process. We could look into different options for representing the arrays.
2. If that fails, as @ngreenwald, we could return a smaller subsetted array. The median will not be as precise but still acceptable.

### LINKED ISSUE #62: Incorrect implementation of `_minimum_larger_value_in_sorted` when `val` matches value in `low_range`
**Describe the bug**

The function `_minimum_larger_value_in_sorted` causes the extracted image signal to be underreported when an extracted `time` value matches a lower bound in `low_range`.

This is because the binary search implementation does not contain a case to handle `if low_range[mid] == val`. Thus, it falls through to the `else` case and creates an off-by-one error (it checks the bound corresponding to the mass below what should be matched).

**Expected behavior**

When a `time` value matches a value in `low_range`, it should immediately return the corresponding index. This will allow the correct signal to be reported (and intensities, widths, and pulse counts for the histograms function).

**To Reproduce**

Run extraction as is.

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Closes #58. Closes #62. Optimizes extraction of histograms needed for median pulse height calculations by doing it in parallel.

**How did you implement your changes**

Modifies:

- `bin_files.get_histograms_per_tof`
- `bin_files.get_median_pulse_height`
- `_extract_bin.c_extract_histograms`
- `_extract_bin.extract_histograms`

The functions in `bin_files.py` now take a list of channels (or `None` to extract data for all channels). This then propagates over to `c_extract_histograms` so the underlying loops only need to run once (previously, they ran multiple times for each channel, needlessly increasing the complexity by a factor of num_channels_to_extract).

**Remaining issues**

This will need a new release, but we'll need to keep `toffy` pinned at `mibi-bin-tools==0.2.10` for the time-being, since this introduces breaking changes. A separate PR will update `toffy` to be compatible with the new median pulse height calculation process (and subsequence `0.2.11` release).

### CODE REVIEW

File: tests/bin_files_test.py [hunk 1]
```diff
@@ -279,7 +279,7 @@ def test_get_width_histogram(test_dir, fov, panel):
     bin_files.get_histograms_per_tof(
         test_dir,
         fov['json'].split('.')[0],
-        'SMA',
+        ['SMA'],
```
Comment (srivarra) [inline tests/bin_files_test.py:282]: Can we add a test for multiple channels selected?

**Review by srivarra — CHANGES_REQUESTED** [review]
Looks good, just a suggestion for an additional test.

**Review by srivarra — APPROVED** [review]
Looks good!

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Did you confirm output is the same with this version and previous version for real data?

**Comment by ngreenwald** [comment]
Maybe first leave the old, incorrect logic for corner cases intact, and make sure the parallelization doesn't change anything. Once you've confirmed that, see how big of a difference switching the logic makes across some real data

**Comment by alex-l-kong** [comment]
@ngreenwald already been testing on some SPAIN cohort FOVs, confirmed that the outputs (widths, intensities, pulse_counts, and median pulse heights) match up for all of them. Optimization provides a 5-7x speed increase.

**Review by camisowers — APPROVED** [review]
Code looks good to me.

**Comment by alex-l-kong** [comment]
> Okay got it. And how significantly do the outputs change with the fixed logic?

With the updated `_minimum_larger_value_in_sorted`, only around 0.05% of pixels are affected (makes sense, since #62 would only apply to a small subset of pixels). However, of those affected, their intensities change dramatically, up to a thousand or even 5000 or higher in some cases.

**Comment by ngreenwald** [comment]
Okay, so the overall MPH of a given image doesn't change appreciably?

**Comment by alex-l-kong** [comment]
> Okay, so the overall MPH of a given image doesn't change appreciably?

Barely any changes at all, single digit differences at most.

**Review by ngreenwald — APPROVED** [review]

**Pull request merged**